### DISCUSSION TITLE
Error [ERR_MODULE_NOT_FOUND] importing { Metaplex } from "@metaplex-foundation/js-next"

### DISCUSSION THREAD

**Discussion by katewayo** [metaplex-foundation/js#63] [Q&A] [answered]
Hi! Hope all is well. I'm building a node + express server application that's trying to call the new API. I don't have any issues when importing things like @solana/web3.js and bip39, but unfortunately I'm getting an error at node initialization when importing Metaplex. 

Here's the error:

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/kirtd/thesyncfloor/apps/hermes/node_modules/@metaplex-foundation/js-next/dist/esm/index.js' imported from /Users/kirtd/thesyncfloor/apps/hermes/solana.js**

And here's the import line:

import { Metaplex } from "@metaplex-foundation/js-next";

Any ideas?

Thanks in advance,

Kirt.
enderzshadow.sol
@katewayo on Twitter

**Comment by lorisleiva** [comment]
Hey 👋  Thanks for raising this. We've tested with ESM on the browser and CJS with Node.js but I don't think we've tried ESM with Node.js yet. What version of the SDK are you using? 🙂

**Reply by katewayo** [reply]
Ah! Here's the version: @metaplex-foundation/js-next@0.6.0

And it it helps, this is the relevant object in package-lock.json:

    "@metaplex-foundation/js-next": {
      "version": "github:metaplex-foundation/js-next#60653007fac0525652837b07e08de890bd64d0c9",
      "from": "github:metaplex-foundation/js-next",
      "requires": {
        "@aws-sdk/client-s3": "^3.54.1",
        "@bundlr-network/client": "^0.7.1",
        "@metaplex-foundation/beet": "^0.1.0",
        "@metaplex-foundation/beet-solana": "^0.1.1",
        "@metaplex-foundation/mpl-token-metadata": "^2.1.0",
        "@solana/spl-token": "^0.2.0",
        "@solana/wallet-adapter-base": "^0.9.3",
        "@solana/web3.js": "^1.37.0",
        "abort-controller": "^3.0.0",
        "bignumber.js": "^9.0.2",
        "bn.js": "^5.2.0",
        "bs58": "^5.0.0",
        "buffer": "^6.0.3",
        "cross-fetch": "^3.1.5",
        "eventemitter3": "^4.0.7",
        "lodash.clonedeep": "^4.5.0",
        "mime": "^3.0.0",
        "tweetnacl": "^1.0.3"
      },

**Reply by lorisleiva** [reply]
Thanks for that. I'm not sure what could go wrong with this setup. Would you be able to create a public repo with the minimum step required to reproduce the issue? That way I can clone and debug. :)

**Reply by katewayo** [reply]
Hey Loris. I did a very simple test node express server (using the express sample) with a metaplex import and it hits the same issue. All you need to do is create a test directory, npm init, npm install express, npm install @metaplex-foundation/js-next, and add the file contents below for index.js and package.json. Upon running node index.js, you'll see the same error. My node version is 16.4.2. Let me know if that gets you the right setup to debug or if you need anything else. 

index.js
====
const express = require('express')
const app = express()
const port = 3000

import { Metaplex } from '@metaplex-foundation/js-next';

app.get('/', (req, res) => {
  res.send('Hello World!')
})

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})

package.json
====
{
  "name": "test",
  "version": "1.0.0",
  "description": "Test",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@metaplex-foundation/js-next": "^0.6.0",
    "express": "^4.17.3"
  }
}

**Reply by lorisleiva** [reply]
Thanks! I'll get started on that today. ~~I'll transform this into an issue as it's likely a bug.~~

**EDIT**: Actually I think the issue is to do with how we import Metaplex on an express app so I'll add a new answer here instead.

**Comment by lorisleiva** [comment, marked as answer]
Thanks again for providing the steps to reproduce this issue. I've been able to make it work using an express app.

Since express apps use "CommonJS", we need to use `require` to import libraries (as you do to import `express` in the first place). That means you need to make the following change in your `index.js` file and it will work as expected.

```diff
- import { Metaplex } from '@metaplex-foundation/js-next';
+ const { Metaplex } = require('@metaplex-foundation/js-next');
```

Here's a quick example of how to fetch an NFT using its mint address on an express app. Note that you'll first need to install web3.js using `npm install @solana/web3.js`.

```js
const express = require('express')
const app = express()
const port = 3000
const { Connection, clusterApiUrl, PublicKey} = require('@solana/web3.js');
const { Metaplex } = require('@metaplex-foundation/js-next');
const mx = Metaplex.make(new Connection(clusterApiUrl('devnet')));

app.get('/', (req, res) => {
    res.send('Hello World!')
})

app.listen(port, () => {
    const mint = new PublicKey("K9iFUvDYK1wsN2ERSU65jUE3mrBesr5UZvrjsvPozj3");
    mx.nfts().findNftByMint(mint).then(console.log);
})
```

**Reply by katewayo** [reply]
Hi Loris! Thank you for investigating. That definitely was part of the problem, but once I made those changes I was still having the same issue. 

As it turns out, the metaplex package I installed via npm was being pulled from my local git clone (I just use that to conveniently look at the source). I happened to notice that the resolution data in package-lock.json wasn't pointing to the official registry 8-|. 

Not sure why that was happening, but once I made the changes you suggested plus uninstalled and reinstalled the metaplex package via npm, the example worked! Thanks again!!

**Reply by lorisleiva** [reply]
Oh, that's a weird one haha. Thanks for sharing, I'm sure this will help others in the same situation. 🍺
